src/java.base/share/classes/java/util/concurrent/LinkedTransferQueue.java

```diff
@@ -307,151 +368,114 @@ final DualNode cmpExNext(DualNode cmp, DualNode val) {
+            long deadline = (timed) ? System.nanoTime() + ns : 0L;
+            boolean upc = isUniprocessor;  // don't spin but later recheck
             Thread w = Thread.currentThread();
-            boolean canSpin = (!w.isVirtual() && spin);
-            int spins = (canSpin && !isUniprocessor) ? SPINS : 0;
-            Object match;
-            while ((match = item) == e && --spins >= 0)
-                Thread.onSpinWait();
-            if (match == e) {
-                boolean uni;
-                if (canSpin && isUniprocessor !=
-                    (uni = (Runtime.getRuntime().availableProcessors() <= 1)))
-                    isUniprocessor = uni; // reset for next time
-                LockSupport.setCurrentBlocker(blocker);
-                waiter = w;
-                VarHandle.fullFence(); // ensure ordering
-                while ((match = item) == e) {
-                    long ns = 0L;
-                    if (w.isInterrupted() ||
-                        (timed && ((ns = deadline - System.nanoTime()) <= 0L))) {
-                        match = cmpExItem(e, (e == null) ? this : null);
-                        break;         // try to cancel with impossible match
+            if (w.isVirtual())             // don't spin
+                spin = false;
+            int spins = (spin & !upc) ? SPINS : 0; // negative when may park
+            while ((m = item) == e) {
+                if (spins >= 0) {
+                    if (--spins >= 0)
+                        Thread.onSpinWait();
+                    else {                 // prepare to park
+                        if (spin)          // occasionally recheck
+                            checkForUniprocessor(upc);
+                        LockSupport.setCurrentBlocker(blocker);
+                        waiter = w;        // ensure ordering
+                        VarHandle.fullFence();
                     }
-                    if (timed) {
-                        if (ns < SPIN_FOR_TIMEOUT_THRESHOLD)
-                            Thread.onSpinWait();
-                        else
-                            LockSupport.parkNanos(ns);
-                    } else if (w instanceof ForkJoinWorkerThread) {
-                        try {
-                            ForkJoinPool.managedBlock(this);
-                        } catch (InterruptedException cannotHappen) { }
-                    } else
-                        LockSupport.park();
-                }
+                } else if (w.isInterrupted() ||
+                           (timed &&       // try to cancel with impossible match
+                            ((ns = deadline - System.nanoTime()) <= 0L))) {
+                    m = cmpExItem(e, (e == null) ? this : null);
+                    break;
+                } else if (timed) {
+                    if (ns < SPIN_FOR_TIMEOUT_THRESHOLD)
+                        Thread.onSpinWait();
+                    else
+                        LockSupport.parkNanos(ns);
```

Comment:
@DougLea If `ns` is sufficiently large, would it not make sense to use managed blocking in case the current thread is a FJWT as below? 🤔

Reply:
It's a good point, but we don't normally do this. Added (L336):
     * returns just barely too soon. As is the case in most j.u.c
     * blocking support, untimed waits use ManagedBlockers when
     * callers are ForkJoin threads, but timed waits use plain
     * parkNanos, under the rationale that known-to-be transient
     * blocking doesn't require compensation. (This decision should be
     * revisited here and elsewhere to deal with very long timeouts.)

Reply:
I should have mentioned that the tradeoffs include that using ManagedBlocking for timed waits here adds overhead: -- constructing of another object just to track the time, and duplicate checks of other parts of await(. Which requires choosing a threshold where it would be worthwhile. Which is  a hard-to-win decision.

Reply:
@DougLea I suspect that a good "governing principle" here is: is the timed park expected to be prematurely unparked, and is the delay sufficiently large to warrant compensation—use ManagedBlocker. Ideally (perhaps) the call to ManagedBlocker takes an "expected conservative estimate of blocking"-nanos, and then that behavior could be completely enclosed within the ManagedBlocker-construct itself (and thus be uniformly applied). It could also use heuristics about expected lifetime of compensating threads to see if it is justified to add one... Just thoughts off of the top of my head. 🤔